File: scripts/algorithms/M/Maximum Number of NonOverlapping Subarrays With Sum Equals Target/Maximum Number of Non-Overlapping Subarrays With Sum Equals Target.java
// Runtime: 105 ms (Top 5.11%) | Memory: 85 MB (Top 62.04%)
class Solution {
    public int maxNonOverlapping(int[] nums, int target) {
        Map<Integer, Integer> valToPos = new HashMap<>();
        int sums = 0;
        int count = 0;
        int lastEndPos = 0;
        valToPos.put(0, 0);
        for (int i = 0; i < nums.length; i++) {
            sums += nums[i];
            int pos = valToPos.getOrDefault(sums - target, -1);
            if (pos >= lastEndPos) {
                count += 1;
                lastEndPos = i + 1;
            }
            valToPos.put(sums, i + 1);
        }
        return count;
    }
}

File: scripts/algorithms/M/Maximum Number of NonOverlapping Subarrays With Sum Equals Target/Maximum Number of Non-Overlapping Subarrays With Sum Equals Target.js
// Runtime: 196 ms (Top 36.00%) | Memory: 56.5 MB (Top 60.00%)
var maxNonOverlapping = function(nums, target) {
    const seen = new Set();
    let total = 0, result = 0;

    for(let n of nums) {
        total += n;

        if(total === target || seen.has(total - target)) {
            total = 0;
            result++;
            seen.clear()
        } else seen.add(total)
    }
    return result;
};